internal/platform/dir.go

```diff
@@ -0,0 +1,38 @@
+package platform
+
+import (
+	"io/fs"
+	"syscall"
+)
+
+// readdirnamesFile is implemented by os.File in dir.go
+// Note: we use this until we finalize our own FD-scoped file.
+type readdirnamesFile interface {
+	Readdirnames(n int) (names []string, err error)
+}
+
+// Readdirnames is like the function on os.File, but for fs.File. This returns
+// syscall.ENOTDIR if not a directory or syscall.EBADF if closed.
+func Readdirnames(f fs.File, n int) (names []string, err error) {
+	switch f := f.(type) {
+	case readdirnamesFile:
+		names, err = f.Readdirnames(n)
+	case fs.ReadDirFile:
+		entries, err := f.ReadDir(n)
```

Comment:
I think err is shadowed in this block, so non-nil error here won't be captured?

Reply:
great eye

Reply:
great eye. fixed and added a test.